zson/marshal.go

```diff
@@ -519,61 +519,42 @@ func (m *MarshalZNGContext) encodeArrayBytes(arrayVal reflect.Value) (zed.Type,
 }
 
 func (m *MarshalZNGContext) encodeArray(arrayVal reflect.Value) (zed.Type, error) {
-	push := m.Builder
-	m.Builder = *zcode.NewBuilder()
 	m.Builder.BeginContainer()
 	arrayLen := arrayVal.Len()
 	types := make([]zed.Type, 0, arrayLen)
-	typeMap := make(map[zed.Type]struct{})
 	for i := 0; i < arrayLen; i++ {
 		item := arrayVal.Index(i)
 		typ, err := m.encodeValue(item)
 		if err != nil {
 			return nil, err
 		}
 		types = append(types, typ)
-		typeMap[typ] = struct{}{}
 	}
-	m.Builder.EndContainer()
+	uniqueTypes := zed.UniqueTypes(append([]zed.Type(nil), types...))
```

Comment:
This should use slices.Clone (once that is merged) no?

Reply:
Indeed. Done in d5396331acdc54f42851c46793565bbad8f8ca10.